### PR TITLE
MAKE-996: use contexts in connection handlers

### PR DESCRIPTION
**Pull request by Kimchelly**
JIRA: https://jira.mongodb.org/browse/MAKE-996

* Write signal handler for SIGTERM/SIGINT on service.
* Each goroutine servicing a connection should be able to cancel all connection reads/writes when the service `Run()` context is cancelled.
* Each request from should be handled asynchronously to allow multiple concurrent requests from a given connection.

### CODE REVIEW

File: mongowire/sock.go [hunk 1]
```diff
@@ -36,14 +54,20 @@ func ReadMessage(reader io.Reader) (Message, error) {
...
 
 	for read := 0; int32(read) < header.Size-4; {
-		n, err := reader.Read(restBuf[read:])
-		if err != nil {
-			return nil, errors.WithStack(err)
+		readFinished = make(chan readResult)
+		go doRead(readFinished, restBuf)
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
```
Comment (tychoish) [inline mongowire/sock.go]: withstack?

File: mongowire/sock.go [hunk 2]
```diff
@@ -36,14 +54,20 @@ func ReadMessage(reader io.Reader) (Message, error) {
 	restBuf := make([]byte, header.Size-4)
 
 	for read := 0; int32(read) < header.Size-4; {
-		n, err := reader.Read(restBuf[read:])
-		if err != nil {
-			return nil, errors.WithStack(err)
+		readFinished = make(chan readResult)
+		go doRead(readFinished, restBuf)
```
Comment (tychoish) [inline mongowire/sock.go]: should we pass the context to this?

File: mongowire/sock.go [hunk 3]
```diff
@@ -56,19 +80,31 @@ func ReadMessage(reader io.Reader) (Message, error) {
...
+	}
 	for {
-		written, err := writer.Write(buf)
-		if err != nil {
-			return errors.Wrap(err, "error writing to client")
+		writeFinished := make(chan writeRes)
+		go func() {
+			defer close(writeFinished)
+			n, err := writer.Write(buf)
+			writeFinished <- writeRes{n: n, err: err}
```
Comment (tychoish) [inline mongowire/sock.go]: do this against the context?

File: service.go [hunk 4]
```diff
@@ -61,7 +69,7 @@ func (s *Service) Run(ctx context.Context) error {
 
 func writeErrorReply(w io.Writer, err error) error {
 	responseNotOk := birch.EC.Int32("ok", 0)
```
Comment (tychoish) [inline service.go]: seems like this should be -1

File: service.go [hunk 5]
```diff
@@ -42,6 +47,9 @@ func (s *Service) Run(ctx context.Context) error {
 	}
 	defer l.Close()
 
+	ctx, cancel := context.WithCancel(ctx)
+	go handleSignals(ctx, cancel)
```
Comment (tychoish) [inline service.go]: I would let applications do their own signal handling/context handling.

File: service.go [hunk 6]
```diff
@@ -61,7 +69,7 @@ func (s *Service) Run(ctx context.Context) error {
 
 func writeErrorReply(w io.Writer, err error) error {
 	responseNotOk := birch.EC.Int32("ok", 0)
-	errorDoc := birch.EC.String("error", err.Error())
+	errorDoc := birch.EC.String("errmsg", err.Error())
```
Comment (tychoish) [inline service.go]: should we errcheck this?

File: mongowire/sock.go (hunk 2 above)
Comment (Kimchelly) [reply mongowire/sock.go]: Done.

File: mongowire/sock.go (hunk 1 above)
Comment (Kimchelly) [reply mongowire/sock.go]: Done.

File: mongowire/sock.go (hunk 3 above)
Comment (Kimchelly) [reply mongowire/sock.go]: Done.

File: service.go (hunk 5 above)
Comment (Kimchelly) [reply service.go]: Removed.

File: service.go (hunk 4 above)
Comment (Kimchelly) [reply service.go]: The mongo shell expects ok to be 0 if it's not ok and 1 if it is ok to let them write `if !ok` in JavaScript.

File: service.go (hunk 6 above)
Comment (Kimchelly) [reply service.go]: This doesn't expect a nil error, but I added a nil check anyways.

File: service.go [hunk 7]
```diff
@@ -68,10 +62,11 @@ func (s *Service) Run(ctx context.Context) error {
 }
 
 func writeErrorReply(w io.Writer, err error) error {
-	responseNotOk := birch.EC.Int32("ok", 0)
-	errorDoc := birch.EC.String("errmsg", err.Error())
-	doc := birch.NewDocument(responseNotOk, errorDoc)
-
+	responseNotOK := birch.EC.Int32("ok", 0)
```
Comment (tychoish) [inline service.go]: shouldn't this be 1 if err == nil?

Comment (Kimchelly) [reply service.go]: The errmsg is entirely optional. ok 0 means that either there was an error (errmsg) or operation is not supported/not relevant (not considered an error).

File: mongowire/sock.go [hunk 8]
```diff
@@ -1,22 +1,43 @@
...
+		defer close(readFinished)
+		n, err := reader.Read(out)
+		select {
+		case readFinished <- readResult{n: n, err: err}:
+		case <-ctx.Done():
+		}
+	}
+
+	readFinished := make(chan readResult)
+	go doRead(ctx, readFinished, sizeBuf)
```
Comment (tychoish) [inline mongowire/sock.go]: I'd just call this inline rather than calling it as a function.

Comment (Kimchelly) [reply mongowire/sock.go]: Copy-pasted `doRead` into the two places it's used.

**Pull request merged**